packages/mdc-top-app-bar/component.ts

```diff
@@ -57,19 +61,49 @@ export class MDCTopAppBar extends MDCComponent<MDCTopAppBarBaseFoundation> {
     this.scrollTarget_ = window;
   }
 
+  initialSyncWithDOM() {
+    this.handleNavigationIconClick_ = this.foundation_.handleNavigationClick.bind(this.foundation_);
+    this.handleResize_ = this.foundation_.handleResize && this.foundation_.handleResize.bind(this.foundation_);
```

Comment:
This needs to be even more explicite by detecting the type of TopAppBar before assigning callbacks.

Currently it is checking for existence of methods in foundation (for example, `handleResize`), instead it can detect presense of class name similar to how we detect the type of top app bar before instantiating appropriate TopAppBar foundation in `getDefaultFoundation` method. This makes easier for us to discover which TopAppBar variant handles respective events.

Reply:
Having the detection of class names in the component is business logic. I thought we were trying to avoid more business logic in the component layer. The foundations set/unset the handleScroll and handleResize value (function or undefined). There is a bit a magic that is happening, but I believe the foundation should work this way.

Reply:
This business logic is very specific to component that determines appropriate foundation to initialize. See `mdc-drawer/component.ts` for example.

We can avoid having this logic in component by creating separate Class for each component variant, but we didn't want to go that route yet.

Reply:
I believe it will look like this:
```
  initialSyncWithDOM() {
    const isFixed = this.root_.classList.contains(cssClasses.FIXED_CLASS);
    const isShort = this.root_.classList.contains(cssClasses.SHORT_CLASS);

    this.handleNavigationClick_ = this.foundation_.handleNavigationClick.bind(this.foundation_);
    this.handleWindowResize_
      = this.foundation_.handleWindowResize && this.foundation_.handleWindowResize.bind(this.foundation_);
    this.handleTargetScroll_
      = this.foundation_.handleTargetScroll && this.foundation_.handleTargetScroll.bind(this.foundation_);

    this.scrollTarget_.addEventListener('scroll', this.handleTargetScroll_ as EventListener);

    if (this.navIcon_) {
      this.navIcon_.addEventListener('click', this.handleNavigationClick_ as EventListener);
    }

    if (!isShort && !isFixed) {
      window.addEventListener('resize', this.handleWindowResize_ as EventListener);
    }
  }

```

Does that look ok to you?

Reply:
We might not need this condition if we provide abstract method in base Class of TopAppBar.

```ts
this.foundation_.handleWindowResize && ...
```

It'll always return a Function type.

Reply:
You mean the base class should effectively be a noop?

So we can just do:
```
  initialSyncWithDOM() {
          window.addEventListener('resize', this.handleWindowResize_ as EventListener);
          this.navIcon_.addEventListener('click', this.handleNavigationClick_ as EventListener);
          this.scrollTarget_.addEventListener('scroll', this.handleTargetScroll_ as EventListener);
  }
```

Reply:
Yes.

Reply:
That seems inefficient. We could potentially have 3 event listeners doing nothing.

Reply:
Not if we debounce window `resize` & `scroll` events. But the way foundations implemented is that it assumes component attaches event listeners to specific TopAppBar variants. So, we might've to bind these events to only specific variants.

This logic could potentially be moved to Foundation but it is out of scope of this PR.

Reply:
`So, we might've to bind these events to only specific variants.` --> You mean as it is right now? If not, what does this look like? 

Even if we do debounce, we're still creating noop eventListeners, which can be avoided by keeping this as is.